frontend/src/components/student/AllActivities/ExpeditionTask/ExpeditionWrapper/ExpeditionWrapper.js

```diff
@@ -0,0 +1,93 @@
+import { useEffect, useState, useCallback } from 'react'
+import { useLocation, useNavigate } from 'react-router-dom'
+import { Spinner } from 'react-bootstrap'
+import { EXPEDITION_STATUS } from './ExpeditionWrapperHelpers'
+import { StudentRoutes } from '../../../../../routes/PageRoutes'
+import ExpeditionService from '../../../../../services/expedition.service'
+import QuestionSelectionDoor from '../QuestionSelectionDoor/QuestionSelectionDoor'
+import QuestionAndOptions from '../QuestionAndOptions/QuestionAndOptions'
+import Timer from '../Timer/Timer'
+// wrapped elements should be:
+
+// -- in ANSWER
+// QuestionAndOptions
+// ClosedQuestionPage
+// OpenQuestionPage
+
+// -- in CHOOSE
+// QuestionSelectionDoor
+
+export function ExpeditionWrapper() {
+  const navigate = useNavigate()
+  const location = useLocation()
+  const { activityId, alreadyStarted } = location.state
+
+  const [expeditionState, setExpeditionState] = useState(undefined)
+
+  const goToSummary = () => {
+    navigate(StudentRoutes.GAME_MAP.GRAPH_TASK.SUMMARY, {
+      state: { expeditionId: activityId, remainingTime: expeditionState.timeRemaining }
+    })
+  }
+
+  // we will pass this function to "lower" components so that we can reload info from endpoint
+  // in wrapper on changes
+  // if it breaks, check whether we don't need to pass activityId in here explicitly
+  const reloadState = useCallback(
+    () =>
+      ExpeditionService.getCurrentState(activityId).then((response) => {
+        setExpeditionState(response)
+      }),
+    [activityId]
+  )
+
+  useEffect(() => {
+    if (activityId) {
+      if (alreadyStarted) {
+        // no need to start the expedition again, just do getInfo
+        reloadState()
+      } else {
+        // else start the expedition and do getInfo then
+        ExpeditionService.setExpeditionStart(activityId).then(() => {
+          reloadState()
+        })
+      }
+    }
+  }, [activityId, alreadyStarted, reloadState])
+
+  if (expeditionState === undefined) {
+    return <Spinner />
+  }
+
+  if (expeditionState.remainingTime <= 0) {
+    goToSummary()
+  }
```

Comment:
Wrzucił bym to w useEffect zależny od remainingTime. Nie może ten if tak sobie leżeć bo jest sprawdzany przy zmianie czegokolwiek, a to bez sensu.